### PR TITLE
lftp 4.7.4

### PR DESCRIPTION
**Pull request by mazgi**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Add modified version for OSX.

References:

* https://github.com/lavv17/lftp/commit/f38a17745ef01e1bca9a3420995e193bf9768c7c
* https://github.com/Homebrew/brew/issues/1387
* https://github.com/Homebrew/homebrew-boneyard/blob/c4daf887cf6ef32a7c197edcbe362f8b239046ed/lftp.rb#L4

### CODE REVIEW

**Comment by ilovezfs** [comment]
Cool! Glad this is getting macOS compatibility without patching again. Any chance there's a mirror (https) link too that you can find?

**Comment by dunn** [comment]
```
Error: 1 problem in 1 formula
lftp:
  * lftp seems to be listed in tap_migrations.json!
    Please remove lftp from present tap & tap_migrations.json
    before submitting it to Homebrew/homebrew-core.
```
Mind taking care of that? Thanks so much!

**Comment by mazgi** [comment]
> a mirror (https) link

I think [GitHub release page](https://github.com/lavv17/lftp/releases/tag/v4.7.4) instead of the `https://lftp.yar.ru/ftp/*`.

**Comment by ilovezfs** [comment]
I'm hoping for the canonical link + a mirror assuming the checksums match. If the GitHub checksum matches the upstream tarball, cool. If not, maybe upstream can post it on GitHub, or maybe there's a mirror elsewhere.

**Comment by ilovezfs** [comment]
Got an error on macOS 10.10 Yosemite: https://bot.brew.sh/job/Homebrew%20Core/12279/version=yosemite/consoleText
```
/bin/sh ../libtool --silent  --tag=CXX   --mode=compile clang++ -DHAVE_CONFIG_H -I. -I../lib  -I../lib -I../trio -I/usr/include -I/usr/local/opt/openssl/include    -O2 -Wall -Wwrite-strings -Woverloaded-virtual -fno-exceptions -fno-rtti -fno-implement-inlines -c -o liblftp_network_la-lftp_ssl.lo `test -f 'lftp_ssl.cc' || echo './'`lftp_ssl.cc
/bin/sh ../libtool --silent  --tag=CXX   --mode=compile clang++ -DHAVE_CONFIG_H -I. -I../lib  -I../lib -I../trio -I/usr/include -I/usr/local/opt/openssl/include    -O2 -Wall -Wwrite-strings -Woverloaded-virtual -fno-exceptions -fno-rtti -fno-implement-inlines -c -o liblftp_network_la-buffer_ssl.lo `test -f 'buffer_ssl.cc' || echo './'`buffer_ssl.cc
/bin/sh ../libtool --silent  --tag=CXX   --mode=compile clang++ -DHAVE_CONFIG_H -I. -I../lib  -I../lib -I../trio -I/usr/include -I/usr/local/opt/openssl/include    -O2 -Wall -Wwrite-strings -Woverloaded-virtual -fno-exceptions -fno-rtti -fno-implement-inlines -c -o liblftp_network_la-RateLimit.lo `test -f 'RateLimit.cc' || echo './'`RateLimit.cc
lftp_ssl.cc:827:14: error: use of undeclared identifier 'SSL_OP_NO_TLSv1_1'
         {"-TLS1.1",SSL_OP_NO_TLSv1_1},
                    ^
lftp_ssl.cc:828:14: error: use of undeclared identifier 'SSL_OP_NO_TLSv1_2'
         {"-TLS1.2",SSL_OP_NO_TLSv1_2},
                    ^
2 errors generated.
make[2]: *** [liblftp_network_la-lftp_ssl.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [install-recursive] Error 1
make: *** [install] Error 2
```
So either it will need to be restricted to >= 10.11, or that error will need to be fixed.

**Comment by mazgi** [comment]
Oops... 😱 

How can I specify a dependency version?
I referred to this: https://github.com/Homebrew/homebrew-core/blob/2e4c18bba2769ca8c383c6565e2a924d0dde31b5/Formula/predictionio.rb#L13

```
Error: 1 problem in 1 formula
lftp:
  * Dependency openssl does not define option "1.0.1+"
==> brew bottle --verbose --json lftp
==> FAILED
```

File: Formula/lftp.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Lftp < Formula
+  desc "Sophisticated file transfer program"
+  homepage "https://lftp.yar.ru/"
+  url "https://lftp.yar.ru/ftp/lftp-4.7.4.tar.xz"
+  sha256 "bf67c4d128b6f769a4082947376a9679c5ee3463a24ab761a0757f75d70bd92c"
+
+  depends_on "pkg-config" => :build
+  depends_on "readline"
+  depends_on :openssl => "1.0"
```
Comment (MikeMcQuaid) [inline Formula/lftp.rb]: This should be `depends_on "openssl"`

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
> How can I specify a dependency version?

You can't.

File: Formula/lftp.rb (hunk 1 above)
Comment (mazgi) [inline Formula/lftp.rb]: Thank you!

**Comment by MikeMcQuaid** [comment]
> So either it will need to be restricted to >= 10.11, or that error will need to be fixed.

Back to that error.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by mazgi** [comment]
>> So either it will need to be restricted to >= 10.11, or that error will need to be fixed.

> Back to that error.

Yes.
I'll try avoid this error.

**Comment by ilovezfs** [comment]
@mazgi you also need to add
```
                          "--with-readline=#{Formula["readline"].opt_prefix}"
```
since it's currently printing
```
checking for Readline headers in /usr/include/readline... found
checking for Readline libraries... found
checking for Readline... yes
checking if Readline version is >= 5.0... expr: syntax error
expr: syntax error
no
configure: WARNING: Found Readline ., which is older than required. Possible compilation failure.
```
meaning it's not picking up your `depends_on "readline"`

**Comment by ilovezfs** [comment]
Also, it prints
```
configure: WARNING: Libidn not found
checking if Libidn should be used... no
```
Brew has libidn, so I'd suggest adding `depends_on "libidn" and also "--with-libidn=#{Formula["libidn"].opt_prefix}" just to be safe.

**Comment by dzg** [comment]
Is there any way to install `lftp` on Mac beyond version 4.6.6 ?

**Comment by dunn** [comment]
@dzg if you want to build on top of this PR and get lftp compiling on all platforms, we'll have it in Homebrew again!

**Comment by dzg** [comment]
@dunn 

> if you want to build on top of this PR and get lftp compiling on all platforms, we'll have it in Homebrew again!

Sorry, I'm a noob, I have no idea what you mean.

**Comment by HazCod** [comment]
For anyone interested, latest syntax error:

```
$ brew install https://raw.githubusercontent.com/HazCod/homebrew-boneyard/master/lftp.rb
######################################################################## 100.0%
==> Downloading https://lftp.yar.ru/ftp/lftp-4.7.5.tar.xz
==> Downloading from https://lftp.tech/ftp/lftp-4.7.5.tar.xz
######################################################################## 100.0%
==> ./configure --prefix=/usr/local/Cellar/lftp/4.7.5 --with-openssl=/usr/local/opt/openssl
==> make install
Last 15 lines from /Users/niels/Library/Logs/Homebrew/lftp/02.make:
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/netinet6/in6.h:292:7: note: expanded from macro 'IN6_IS_ADDR_LINKLOCAL'
        (((a)->s6_addr[0] == 0xfe) && (((a)->s6_addr[1] & 0xc0) == 0x80))
          ~~~^ ~~~~~~~
Resolver.cc:321:10: error: member reference base type 'const char' is not a structure or union
      if(IN6_IS_ADDR_LINKLOCAL(address) && scope==0) {
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/netinet6/in6.h:292:37: note: expanded from macro 'IN6_IS_ADDR_LINKLOCAL'
        (((a)->s6_addr[0] == 0xfe) && (((a)->s6_addr[1] & 0xc0) == 0x80))
                                        ~~~^ ~~~~~~~
2 errors generated.
make[2]: *** [liblftp_network_la-Resolver.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [install-recursive] Error 1
make: *** [install] Error 2
```

**Comment by fxcoudert** [comment]
The build failure has a simple patch which has now been committed upstream (https://github.com/lavv17/lftp/commit/259d642e1fea2ddf38763d49e8e7701f0a947d4c).

Here is a version of the formula that includes: the patch, the correct configuration to get readline, and requires and uses libidn: https://github.com/fxcoudert/homebrew-core/commit/45bd30ca09cd02290ecd2bf599c9023ef09744d3

@mazgi can you update your pull request with it?

**Comment by HazCod** [comment]
@fxcoudert Can confirm it works now, thank you!
Can't wait for the PR.

**Comment by mazgi** [comment]
@fxcoudert Wow! Great ❗️ Sure 😄 

I'm trying find your great commit: https://github.com/fxcoudert/homebrew-core/commit/45bd30ca09cd02290ecd2bf599c9023ef09744d3

```shellsession
$ date
Sat Jan 14 05:10:08 JST 2017
$ git remote -v
origin	git@github.com:fxcoudert/homebrew-core.git (fetch)
origin	git@github.com:fxcoudert/homebrew-core.git (push)
$ git fetch origin --prune
$ git merge --ff-only origin master
Already up-to-date.
$ git log -1
commit 7d31ff24008913ad07290ae3c8580c3b07337f14
Author: BrewTestBot <brew-test-bot@googlegroups.com>
Date:   Mon Jan 9 15:15:09 2017 +0000

    libspectre: update 0.2.8_1 bottle.
$ git show 45bd30ca09cd02290ecd2bf599c9023ef09744d3
fatal: bad object 45bd30ca09cd02290ecd2bf599c9023ef09744d3
```

File: Formula/lftp.rb [hunk 2]
```diff
@@ -1,21 +1,41 @@
...
+  url "https://lftp.yar.ru/ftp/lftp-4.7.5.tar.bz2"
+  mirror "ftp://ftp.st.ryukoku.ac.jp/pub/network/ftp/lftp/lftp-4.7.5.tar.bz2"
+  sha256 "90f3cbc827534c3b3a391a2dd8b39cc981ac4991fa24b6f90e2008ccc0a5207d"
 
-  depends_on "pkg-config" => :build
   depends_on "readline"
   depends_on "openssl"
+  depends_on "libidn"
+
+  # Fix a cast issue, patch was merged upstream: https://github.com/lavv17/lftp/pull/307
```
Comment (ilovezfs) [inline Formula/lftp.rb]: Can we apply this with a `patch do` block?

**Comment by fxcoudert** [comment]
There is a build failure on yosemite, but I don't understand it: SSL_OP_NO_TLSv1_1 and SSL_OP_NO_TLSv1_2 are defined in <openssl/ssl.h>, which is apparently correctly found by the configure script. Can someone with a Yosemite machine look into it?

**Comment by ilovezfs** [comment]
It means it's picking up the SDK's openssl headers.

**Comment by fxcoudert** [comment]
It's compiled with `-I. -I../lib  -I../lib -I../trio -I/usr/include -I/usr/local/opt/openssl/include  -I/usr/local/opt/libidn/include` flags passed, clang's behaviour is to ignore the redundant `/usr/local`, and thus the ordered list of header search path is as follows:

> #include <...> search starts here:
 .
 ../lib
 ../trio
 /usr/local/opt/openssl/include
 /usr/local/opt/libidn/include
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.0.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)

So our openssl should come before the system header.

**Comment by ilovezfs** [comment]
apparently not

**Comment by fxcoudert** [comment]
As I said, it's hard to guess without having access to a Yosemite machine. One could try to reorder the `-I` flags in the makefiles:

```
--- src/Makefile.am	2016-07-08 11:10:55.000000000 +0200
+++ src/Makefile.am.fx	2017-01-14 16:03:49.000000000 +0100
@@ -50,7 +50,7 @@ cmd_mirror_la_LDFLAGS = -module -avoid-v
 cmd_sleep_la_LDFLAGS  = -module -avoid-version -rpath $(pkgverlibdir)
 cmd_torrent_la_LDFLAGS= -module -avoid-version -rpath $(pkgverlibdir)
 liblftp_pty_la_LDFLAGS     = -avoid-version -rpath $(pkgverlibdir)
-liblftp_network_la_CPPFLAGS = $(AM_CPPFLAGS) $(ZLIB_CPPFLAGS) $(OPENSSL_CPPFLAGS) $(LIBGNUTLS_CFLAGS)
+liblftp_network_la_CPPFLAGS = $(AM_CPPFLAGS) $(OPENSSL_CPPFLAGS) $(LIBGNUTLS_CFLAGS) $(ZLIB_CPPFLAGS)
 liblftp_network_la_LDFLAGS = -avoid-version -rpath $(pkgverlibdir)
 liblftp_network_la_LIBADD  = $(SOCKSLIBS) $(OPENSSL_LDFLAGS) $(OPENSSL_LIBS) $(LIBGNUTLS_LIBS) $(GNULIB) $(ZLIB_LDFLAGS) $(ZLIB)
 
--- src/Makefile.in	2017-01-01 18:26:10.000000000 +0100
+++ src/Makefile.in.fx	2017-01-14 16:03:58.000000000 +0100
@@ -1756,7 +1756,7 @@ cmd_mirror_la_LDFLAGS = -module -avoid-v
 cmd_sleep_la_LDFLAGS = -module -avoid-version -rpath $(pkgverlibdir)
 cmd_torrent_la_LDFLAGS = -module -avoid-version -rpath $(pkgverlibdir)
 liblftp_pty_la_LDFLAGS = -avoid-version -rpath $(pkgverlibdir)
-liblftp_network_la_CPPFLAGS = $(AM_CPPFLAGS) $(ZLIB_CPPFLAGS) $(OPENSSL_CPPFLAGS) $(LIBGNUTLS_CFLAGS)
+liblftp_network_la_CPPFLAGS = $(AM_CPPFLAGS) $(OPENSSL_CPPFLAGS) $(LIBGNUTLS_CFLAGS) $(ZLIB_CPPFLAGS)
 liblftp_network_la_LDFLAGS = -avoid-version -rpath $(pkgverlibdir)
 liblftp_network_la_LIBADD = $(SOCKSLIBS) $(OPENSSL_LDFLAGS) $(OPENSSL_LIBS) $(LIBGNUTLS_LIBS) $(GNULIB) $(ZLIB_LDFLAGS) $(ZLIB)
 proto_ftp_la_LIBADD = liblftp-network.la
```

**Comment by woodruffw** [comment]
Closing in favor of #9594.

**Pull request closed without merging**